File: apps/askbible-android/core/src/main/kotlin/me/askbible/native_/data/BibleCatalog.kt
package me.askbible.native_.data

/**
 * 圣经目录。书卷号 / id / 中文名 / 章数逐条来自 RN 的 `lib/bible/scripture-books.ts`，
 * 与 iOS 的 `BibleCatalog.swift` 对等双写，一致性由 check:book-catalog 对拍保证。
 * 分组名与配色是本 App 目录页的设计（RN 版同款），按 bookNumber 划分归属。
 */
data class BookRef(
    val number: Int,
    val id: String,
    val nameZh: String,
    val nameEn: String,
    val chapterCount: Int,
) {
    val displayNumber: String get() = number.toString().padStart(2, '0')
}

data class BookGroup(
    /** RN scripture_canon_catalog.json 的 sectionId，三语标题按它查 LocaleTables */
    val id: String,
    val name: String,
    val colorHex: Int,
    val books: List<BookRef>,
)

object BibleCatalog {
    /** OLD_TESTAMENT_MAX_BOOK_NUMBER */
    const val OLD_TESTAMENT_MAX = 39

    val groups: List<BookGroup> = listOf(
    BookGroup(
        id = "canon-torah",
        name = "Torah",
        colorHex = 0x2F6291,
        books = listOf(
        BookRef(1, "GEN", "创世记", "Genesis", 50),
        BookRef(2, "EXO", "出埃及记", "Exodus", 40),
        BookRef(3, "LEV", "利未记", "Leviticus", 27),
        BookRef(4, "NUM", "民数记", "Numbers", 36),
        BookRef(5, "DEU", "申命记", "Deuteronomy", 34),
        ),
    ),
    BookGroup(
        id = "canon-history",
        name = "Historical books",
        colorHex = 0xB4560E,
        books = listOf(
        BookRef(6, "JOS", "约书亚记", "Joshua", 24),
        BookRef(7, "JDG", "士师记", "Judges", 21),
        BookRef(8, "RUT", "路得记", "Ruth", 4),
        BookRef(9, "1SA", "撒母耳记上", "1 Samuel", 31),
        BookRef(10, "2SA", "撒母耳记下", "2 Samuel", 24),
        BookRef(11, "1KI", "列王纪上", "1 Kings", 22),
        BookRef(12, "2KI", "列王纪下", "2 Kings", 25),
        BookRef(13, "1CH", "历代志上", "1 Chronicles", 29),
        BookRef(14, "2CH", "历代志下", "2 Chronicles", 36),
        BookRef(15, "EZR", "以斯拉记", "Ezra", 10),
        BookRef(16, "NEH", "尼希米记", "Nehemiah", 13),
        BookRef(17, "EST", "以斯帖记", "Esther", 10),
        ),
    ),
    BookGroup(
        id = "canon-wisdom",
        name = "Wisdom & poetry",
        colorHex = 0x4F7A3F,
        books = listOf(
        BookRef(18, "JOB", "约伯记", "Job", 42),
        BookRef(19, "PSA", "诗篇", "Psalms", 150),
        BookRef(20, "PRO", "箴言", "Proverbs", 31),
        BookRef(21, "ECC", "传道书", "Ecclesiastes", 12),
        BookRef(22, "SNG", "雅歌", "Song of Solomon", 8),
        ),
    ),
    BookGroup(
        id = "canon-major-prophets",
        name = "Major prophets",
        colorHex = 0x8A4A12,
        books = listOf(
        BookRef(23, "ISA", "以赛亚书", "Isaiah", 66),
        BookRef(24, "JER", "耶利米书", "Jeremiah", 52),
        BookRef(25, "LAM", "耶利米哀歌", "Lamentations", 5),
        BookRef(26, "EZK", "以西结书", "Ezekiel", 48),
        BookRef(27, "DAN", "但以理书", "Daniel", 12),
        ),
    ),
    BookGroup(
        id = "canon-minor-prophets",
        name = "Minor prophets",
        colorHex = 0x6B7A3F,
        books = listOf(
        BookRef(28, "HOS", "何西阿书", "Hosea", 14),
        BookRef(29, "JOL", "约珥书", "Joel", 3),
        BookRef(30, "AMO", "阿摩司书", "Amos", 9),
        BookRef(31, "OBA", "俄巴底亚书", "Obadiah", 1),
        BookRef(32, "JON", "约拿书", "Jonah", 4),
        BookRef(33, "MIC", "弥迦书", "Micah", 7),
        BookRef(34, "NAM", "那鸿书", "Nahum", 3),
        BookRef(35, "HAB", "哈巴谷书", "Habakkuk", 3),
        BookRef(36, "ZEP", "西番雅书", "Zephaniah", 3),
        BookRef(37, "HAG", "哈该书", "Haggai", 2),
        BookRef(38, "ZEC", "撒迦利亚书", "Zechariah", 14),
        BookRef(39, "MAL", "玛拉基书", "Malachi", 4),
        ),
    ),
    BookGroup(
        id = "canon-gospels",
        name = "Gospels",
        colorHex = 0xC1660B,
        books = listOf(
        BookRef(40, "MAT", "马太福音", "Matthew", 28),
        BookRef(41, "MRK", "马可福音", "Mark", 16),
        BookRef(42, "LUK", "路加福音", "Luke", 24),
        BookRef(43, "JHN", "约翰福音", "John", 21),
        ),
    ),
    BookGroup(
        id = "canon-church-history",
        name = "Church history",
        colorHex = 0x12786F,
        books = listOf(
        BookRef(44, "ACT", "使徒行传", "Acts", 28),
        ),
    ),
    BookGroup(
        id = "canon-pauline",
        name = "Pauline epistles",
        colorHex = 0x6D4FC4,
        books = listOf(
        BookRef(45, "ROM", "罗马书", "Romans", 16),
        BookRef(46, "1CO", "哥林多前书", "1 Corinthians", 16),
        BookRef(47, "2CO", "哥林多后书", "2 Corinthians", 13),
        BookRef(48, "GAL", "加拉太书", "Galatians", 6),
        BookRef(49, "EPH", "以弗所书", "Ephesians", 6),
        BookRef(50, "PHP", "腓立比书", "Philippians", 4),
        BookRef(51, "COL", "歌罗西书", "Colossians", 4),
        BookRef(52, "1TH", "帖撒罗尼迦前书", "1 Thessalonians", 5),
        BookRef(53, "2TH", "帖撒罗尼迦后书", "2 Thessalonians", 3),
        BookRef(54, "1TI", "提摩太前书", "1 Timothy", 6),
        BookRef(55, "2TI", "提摩太后书", "2 Timothy", 4),
        BookRef(56, "TIT", "提多书", "Titus", 3),
        BookRef(57, "PHM", "腓利门书", "Philemon", 1),
        ),
    ),
    BookGroup(
        id = "canon-general-epistles",
        name = "General epistles",
        colorHex = 0x9B4B8F,
        books = listOf(
        BookRef(58, "HEB", "希伯来书", "Hebrews", 13),
        BookRef(59, "JAS", "雅各书", "James", 5),
        BookRef(60, "1PE", "彼得前书", "1 Peter", 5),
        BookRef(61, "2PE", "彼得后书", "2 Peter", 3),
        BookRef(62, "1JN", "约翰一书", "1 John", 5),
        BookRef(63, "2JN", "约翰二书", "2 John", 1),
        BookRef(64, "3JN", "约翰三书", "3 John", 1),
        BookRef(65, "JUD", "犹大书", "Jude", 1),
        ),
    ),
    BookGroup(
        id = "canon-apocalyptic",
        name = "Apocalypse",
        colorHex = 0xA8324A,
        books = listOf(
        BookRef(66, "REV", "启示录", "Revelation", 22),
        ),
    ),
    )

    val all: List<BookRef> = groups.flatMap { it.books }

    val oldTestament: List<BookGroup> get() = groups.filter { it.books.first().number <= OLD_TESTAMENT_MAX }
    val newTestament: List<BookGroup> get() = groups.filter { it.books.first().number > OLD_TESTAMENT_MAX }

    fun book(id: String): BookRef? = all.firstOrNull { it.id == id }
}

File: apps/askbible-ios/AskBible/Model/BibleCatalog.swift
import SwiftUI

/// 圣经目录。书卷号 / id / 中文名 / 章数逐条来自 RN 的 `lib/bible/scripture-books.ts`，
/// 与 Android 的 `BibleCatalog.kt` 对等双写，一致性由 check:book-catalog 对拍保证。
/// 分组名与配色是本 App 目录页的设计，按 bookNumber 划分归属。
struct BookRef: Identifiable, Hashable {
    let number: Int
    let id: String
    let nameZh: String
    let nameEn: String
    let chapterCount: Int

    var displayNumber: String { String(format: "%02d", number) }
    /// 旧代码的默认名（英文）；页面上请用 `name(_ locale:)` 按展示语言取
    var name: String { nameEn }
}

struct BookGroup: Identifiable, Hashable {
    /// RN scripture_canon_catalog.json 的 sectionId，三语标题按它查 LocaleTables
    let id: String
    let name: String
    let colorHex: UInt32
    let books: [BookRef]

    var color: Color { Color(rgb: colorHex) }
}

enum BibleCatalog {
    /// OLD_TESTAMENT_MAX_BOOK_NUMBER
    static let oldTestamentMax = 39

    static let groups: [BookGroup] = [
        BookGroup(id: "canon-torah", name: "Torah", colorHex: 0x2F6291, books: [
            BookRef(number: 1, id: "GEN", nameZh: "创世记", nameEn: "Genesis", chapterCount: 50),
            BookRef(number: 2, id: "EXO", nameZh: "出埃及记", nameEn: "Exodus", chapterCount: 40),
            BookRef(number: 3, id: "LEV", nameZh: "利未记", nameEn: "Leviticus", chapterCount: 27),
            BookRef(number: 4, id: "NUM", nameZh: "民数记", nameEn: "Numbers", chapterCount: 36),
            BookRef(number: 5, id: "DEU", nameZh: "申命记", nameEn: "Deuteronomy", chapterCount: 34),
        ]),
        BookGroup(id: "canon-history", name: "Historical books", colorHex: 0xB4560E, books: [
            BookRef(number: 6, id: "JOS", nameZh: "约书亚记", nameEn: "Joshua", chapterCount: 24),
            BookRef(number: 7, id: "JDG", nameZh: "士师记", nameEn: "Judges", chapterCount: 21),
            BookRef(number: 8, id: "RUT", nameZh: "路得记", nameEn: "Ruth", chapterCount: 4),
            BookRef(number: 9, id: "1SA", nameZh: "撒母耳记上", nameEn: "1 Samuel", chapterCount: 31),
            BookRef(number: 10, id: "2SA", nameZh: "撒母耳记下", nameEn: "2 Samuel", chapterCount: 24),
            BookRef(number: 11, id: "1KI", nameZh: "列王纪上", nameEn: "1 Kings", chapterCount: 22),
            BookRef(number: 12, id: "2KI", nameZh: "列王纪下", nameEn: "2 Kings", chapterCount: 25),
            BookRef(number: 13, id: "1CH", nameZh: "历代志上", nameEn: "1 Chronicles", chapterCount: 29),
            BookRef(number: 14, id: "2CH", nameZh: "历代志下", nameEn: "2 Chronicles", chapterCount: 36),
            BookRef(number: 15, id: "EZR", nameZh: "以斯拉记", nameEn: "Ezra", chapterCount: 10),
            BookRef(number: 16, id: "NEH", nameZh: "尼希米记", nameEn: "Nehemiah", chapterCount: 13),
            BookRef(number: 17, id: "EST", nameZh: "以斯帖记", nameEn: "Esther", chapterCount: 10),
        ]),
        BookGroup(id: "canon-wisdom", name: "Wisdom & poetry", colorHex: 0x4F7A3F, books: [
            BookRef(number: 18, id: "JOB", nameZh: "约伯记", nameEn: "Job", chapterCount: 42),
            BookRef(number: 19, id: "PSA", nameZh: "诗篇", nameEn: "Psalms", chapterCount: 150),
            BookRef(number: 20, id: "PRO", nameZh: "箴言", nameEn: "Proverbs", chapterCount: 31),
            BookRef(number: 21, id: "ECC", nameZh: "传道书", nameEn: "Ecclesiastes", chapterCount: 12),
            BookRef(number: 22, id: "SNG", nameZh: "雅歌", nameEn: "Song of Solomon", chapterCount: 8),
        ]),
        BookGroup(id: "canon-major-prophets", name: "Major prophets", colorHex: 0x8A4A12, books: [
            BookRef(number: 23, id: "ISA", nameZh: "以赛亚书", nameEn: "Isaiah", chapterCount: 66),
            BookRef(number: 24, id: "JER", nameZh: "耶利米书", nameEn: "Jeremiah", chapterCount: 52),
            BookRef(number: 25, id: "LAM", nameZh: "耶利米哀歌", nameEn: "Lamentations", chapterCount: 5),
            BookRef(number: 26, id: "EZK", nameZh: "以西结书", nameEn: "Ezekiel", chapterCount: 48),
            BookRef(number: 27, id: "DAN", nameZh: "但以理书", nameEn: "Daniel", chapterCount: 12),
        ]),
        BookGroup(id: "canon-minor-prophets", name: "Minor prophets", colorHex: 0x6B7A3F, books: [
            BookRef(number: 28, id: "HOS", nameZh: "何西阿书", nameEn: "Hosea", chapterCount: 14),
            BookRef(number: 29, id: "JOL", nameZh: "约珥书", nameEn: "Joel", chapterCount: 3),
            BookRef(number: 30, id: "AMO", nameZh: "阿摩司书", nameEn: "Amos", chapterCount: 9),
            BookRef(number: 31, id: "OBA", nameZh: "俄巴底亚书", nameEn: "Obadiah", chapterCount: 1),
            BookRef(number: 32, id: "JON", nameZh: "约拿书", nameEn: "Jonah", chapterCount: 4),
            BookRef(number: 33, id: "MIC", nameZh: "弥迦书", nameEn: "Micah", chapterCount: 7),
            BookRef(number: 34, id: "NAM", nameZh: "那鸿书", nameEn: "Nahum", chapterCount: 3),
            BookRef(number: 35, id: "HAB", nameZh: "哈巴谷书", nameEn: "Habakkuk", chapterCount: 3),
            BookRef(number: 36, id: "ZEP", nameZh: "西番雅书", nameEn: "Zephaniah", chapterCount: 3),
            BookRef(number: 37, id: "HAG", nameZh: "哈该书", nameEn: "Haggai", chapterCount: 2),
            BookRef(number: 38, id: "ZEC", nameZh: "撒迦利亚书", nameEn: "Zechariah", chapterCount: 14),
            BookRef(number: 39, id: "MAL", nameZh: "玛拉基书", nameEn: "Malachi", chapterCount: 4),
        ]),
        BookGroup(id: "canon-gospels", name: "Gospels", colorHex: 0xC1660B, books: [
            BookRef(number: 40, id: "MAT", nameZh: "马太福音", nameEn: "Matthew", chapterCount: 28),
            BookRef(number: 41, id: "MRK", nameZh: "马可福音", nameEn: "Mark", chapterCount: 16),
            BookRef(number: 42, id: "LUK", nameZh: "路加福音", nameEn: "Luke", chapterCount: 24),
            BookRef(number: 43, id: "JHN", nameZh: "约翰福音", nameEn: "John", chapterCount: 21),
        ]),
        BookGroup(id: "canon-church-history", name: "Church history", colorHex: 0x12786F, books: [
            BookRef(number: 44, id: "ACT", nameZh: "使徒行传", nameEn: "Acts", chapterCount: 28),
        ]),
        BookGroup(id: "canon-pauline", name: "Pauline epistles", colorHex: 0x6D4FC4, books: [
            BookRef(number: 45, id: "ROM", nameZh: "罗马书", nameEn: "Romans", chapterCount: 16),
            BookRef(number: 46, id: "1CO", nameZh: "哥林多前书", nameEn: "1 Corinthians", chapterCount: 16),
            BookRef(number: 47, id: "2CO", nameZh: "哥林多后书", nameEn: "2 Corinthians", chapterCount: 13),
            BookRef(number: 48, id: "GAL", nameZh: "加拉太书", nameEn: "Galatians", chapterCount: 6),
            BookRef(number: 49, id: "EPH", nameZh: "以弗所书", nameEn: "Ephesians", chapterCount: 6),
            BookRef(number: 50, id: "PHP", nameZh: "腓立比书", nameEn: "Philippians", chapterCount: 4),
            BookRef(number: 51, id: "COL", nameZh: "歌罗西书", nameEn: "Colossians", chapterCount: 4),
            BookRef(number: 52, id: "1TH", nameZh: "帖撒罗尼迦前书", nameEn: "1 Thessalonians", chapterCount: 5),
            BookRef(number: 53, id: "2TH", nameZh: "帖撒罗尼迦后书", nameEn: "2 Thessalonians", chapterCount: 3),
            BookRef(number: 54, id: "1TI", nameZh: "提摩太前书", nameEn: "1 Timothy", chapterCount: 6),
            BookRef(number: 55, id: "2TI", nameZh: "提摩太后书", nameEn: "2 Timothy", chapterCount: 4),
            BookRef(number: 56, id: "TIT", nameZh: "提多书", nameEn: "Titus", chapterCount: 3),
            BookRef(number: 57, id: "PHM", nameZh: "腓利门书", nameEn: "Philemon", chapterCount: 1),
        ]),
        BookGroup(id: "canon-general-epistles", name: "General epistles", colorHex: 0x9B4B8F, books: [
            BookRef(number: 58, id: "HEB", nameZh: "希伯来书", nameEn: "Hebrews", chapterCount: 13),
            BookRef(number: 59, id: "JAS", nameZh: "雅各书", nameEn: "James", chapterCount: 5),
            BookRef(number: 60, id: "1PE", nameZh: "彼得前书", nameEn: "1 Peter", chapterCount: 5),
            BookRef(number: 61, id: "2PE", nameZh: "彼得后书", nameEn: "2 Peter", chapterCount: 3),
            BookRef(number: 62, id: "1JN", nameZh: "约翰一书", nameEn: "1 John", chapterCount: 5),
            BookRef(number: 63, id: "2JN", nameZh: "约翰二书", nameEn: "2 John", chapterCount: 1),
            BookRef(number: 64, id: "3JN", nameZh: "约翰三书", nameEn: "3 John", chapterCount: 1),
            BookRef(number: 65, id: "JUD", nameZh: "犹大书", nameEn: "Jude", chapterCount: 1),
        ]),
        BookGroup(id: "canon-apocalyptic", name: "Apocalypse", colorHex: 0xA8324A, books: [
            BookRef(number: 66, id: "REV", nameZh: "启示录", nameEn: "Revelation", chapterCount: 22),
        ]),
    ]

    static let all: [BookRef] = groups.flatMap(\.books)

    static var oldTestament: [BookGroup] { groups.filter { ($0.books.first?.number ?? 0) <= oldTestamentMax } }
    static var newTestament: [BookGroup] { groups.filter { ($0.books.first?.number ?? 0) > oldTestamentMax } }

    static func book(id: String) -> BookRef? { all.first { $0.id == id } }
}
